storage/blockdevice/include/blockdevice/HeapBlockDevice.h

```diff
@@ -93,8 +95,6 @@ class HeapBlockDevice : public BlockDevice {
     virtual int read(void *buffer, bd_addr_t addr, bd_size_t size);
 
     /** Program blocks to a block device
-     *
-     *  The blocks must have been erased prior to being programmed
```

Comment:
I removed this because it's definitely not true - the buffer is only allocated when a block is programmed, not when it's erased.

Reply:
I read this as being about checks the class method will do to enforce that the block is erased before being programmed; policy you must follow as a programmer. I don't see code to check this, but maybe there ought to be some code here if we want to ensure programs written to use the HeapBlockDevice will continue to work when used with actual flash-based devices.

I checked the implementation of `is_valid_program()`, but it doesn't appear to check that a block was erased before it can be programmed. If this policy is enforced anywhere, I'm not clear on where yet.

Reply:
Good point. I think it's a real inconsistency in our existing API.

Mbed OS allows a `BlockDevice` to be either flash-based or not.
* For flash-based ones (e.g. `SPIFBlockDevice`), I think the actual erase requirement is set by the hardware (if I understand correctly).
* For non-flash-based ones, "erase" should in theory be unsupported. One example is SD. From [its documentation](https://os.mbed.com/docs/mbed-os/v6.9/apis/sdblockdevice.html)
    > Because an SD card doesn't need preparation, erase is not applicable on SDBlockDevice; it's a no-op operation.

So there's no single policy for all `BlockDevice`s.

`HeapBlockDevice` has always been quite self-contradicting:
* It was _intended_ to be non-flash-based. This was why we used to use a flash simulation layer [`FlashSimBlockDevice`](https://github.com/ARMmbed/mbed-os/blob/master/storage/blockdevice/source/FlashSimBlockDevice.cpp) on top of it, to simulate erase operations when used for `TDBStore` which used to required erase (but not anymore).
* However, a partially implemented & partially lacking `HeapBlockDevice::erase()` always existed - it performs size/alignment check, but doesn't do the actual erase.
* And this line of Doxygen we are talking about.

In my opinion, the proper solution for best consistency is remove `::erase` and any mentions of "erase" from `HeapBlockDevice`. This would change the API for users who do like to use `::erase`.

Any preference?

Reply:
In the SD example I used above, its Doxygen contradicts with it's documentation.
Documentation:
> Because an SD card doesn't need preparation, erase is not applicable on SDBlockDevice; it's a no-op operation.
Doxygen:
https://github.com/ARMmbed/mbed-os/blob/8075b2348cf8250a99f47eb410ad42e1ef2eea6a/storage/blockdevice/COMPONENT_SD/include/SD/SDBlockDevice.h#L122-L122

Reply:
Could you update SDBlockDevice too, please?

If we remove erase from HeapBlockDevice, it might no longer be swappable for other BlockDevice implementations. It may need to be a no-op. Also, If HeapBlockDevice::program() allocates memory, what should free the allocated memory? In an earlier iteration of this PR, you had HeapBlockDevice::erase() free memory. If HeapBlockDevice::erase() is instead a no-op, should we move the freeing of allocated memory to HeapBlockDevice::~HeapBlockDevice() instead?

Reply:
The base class's virtual `BlockDevice::erase()` is already a no-op. The freeing of allocated memory is also done in `HeapBlockDevice::~HeapBlockDevice()` already. So removing `HeapBlockDevice::erase()` and updating Doxygen should be enough.

Reply:
Following discussions below, I kept `HeapBlockDevice::erase()` because it doesn't break compatibility/consistency with the definition of the `BlockDevice` API.